### ISSUE TITLE
Input debug file shows entire config, but web interface shows incomplete config

### ISSUE THREAD

**Issue by thecityofguanyu** [ytti/oxidized#474]
As the title indicates.

First off, versions:

```
root@btcnms:~/.config/oxidized# gem list | grep oxid
oxidized (0.14.1, 0.12.2, 0.9.0)
oxidized-script (0.1.2, 0.1.1)
oxidized-web (0.7.0, 0.5.2)
```

Now the screenshot from a currently affected config in the Oxidized web interface:

![oxidized-web-screenshot](http://i.imgur.com/D4FdnmU.png)

From the Oxidized log:

```
D, [2016-06-19T16:29:13.172496 #24068] DEBUG -- : Jobs 0, Want: 1
D, [2016-06-19T16:29:15.936425 #24068] DEBUG -- : SSH: terminal length 0 @ host1
D, [2016-06-19T16:29:16.542699 #24068] DEBUG -- : SSH: show version @ host1
D, [2016-06-19T16:29:17.153067 #24068] DEBUG -- : SSH: show running-config @ host1
D, [2016-06-19T16:29:17.364247 #24068] DEBUG -- : SSH: exit @ host1
D, [2016-06-19T16:29:17.364427 #24068] DEBUG -- : SSH: exit @ host1
I, [2016-06-19T16:29:19.212219 #24068]  INFO -- : Configuration updated for /host1
D, [2016-06-19T16:29:19.212320 #24068] DEBUG -- : Jobs 0, Want: 1
```

And here's the generated input debug file from that host. 318 lines containing the entire config, as expected.

```
root@oxidized:~/oxidized# wc -l log-10.35.12.10-ssh
318 log-10.35.12.10-ssh
```

I won't post the whole config from the debug, but here's a relevant snippet of it. If you compare this to the screenshot of the web interface, you can see that it appears that Oxidized just sometimes isn't displaying the running config section of that config. I really have no idea why. I would think that it would display everything that got was also logged in the debug log file.

```
[redacted]
Current system image file is "NONVOL:/9212916-2R100906.biz"
Boot system image file is "NONVOL:/9212916-2R100906.biz"
Primary system configuration file is "startup-config"
host1#show running-config
Building configuration...
!
!
! ADTRAN, Inc. OS version R10.9.6.E
[redacted]
```

Again, this happens regularly, but I call it intermittent because I've found no obvious pattern. It's worth nothing that I've only observed this problem using a local model template that I built for Adtran devices. We have 61 devices currently added to Oxidized using that model template. 

With that said, some devices are afflicted and others aren't. And when new configs are downloaded each day, some will correct themselves while other will begin showing only partial configs.

I've pasted the model template below.

``` Ruby
class ADTRAN < Oxidized::Model

  prompt /^([\w.@()-]+[>#]\s?)$/
  comment  '! '

  cmd :all do |cfg|
    cfg.each_line.to_a[1..-2].join
  end

  cmd 'show version' do |cfg|
    comment cfg.lines.first
  end

  cmd 'show running-config' do |cfg|
    cfg = cfg.each_line.to_a[3..-1].join
    cfg
  end

  cfg :telnet do
    username /^Username:/
    password /^Password:/
  end

  cfg :telnet, :ssh do
    post_login 'terminal length 0'
    # preferred way to handle additional passwords
    if vars :enable
      post_login do
        send "enable\n"
        send vars(:enable) + "\n"
      end
    end
    pre_logout 'exit'
    pre_logout 'exit'
  end
end
```

**Comment by laf** [comment]
When we had this issue it was the prompt detection. What does your prompt look like?

**Comment by thecityofguanyu** [comment]
@laf 

Prompt regex from model file:

```
  prompt /^([\w.@()-]+[>#]\s?)$/
```

Except from input debug that contains both privilege and non-privilege level prompts:

```
host1>terminal length 0
host1>enable
Password:
host1#show version
ADTRAN, Inc. OS version R10.9.6.E
  Mainline Version: ENM.13.100
  P4 Changelist: 187892
  Checksum: 934026FD
  Built on: Tue Jan 27 15:02:33 2015
  Upgrade key: 03a91cf2145bb4cb843fc72cedb53a93
Hardware version DSP7119-2ndGen
Boot ROM version 14.04.00
  Checksum: 8549
  Built on: Wed Apr 11 15:24:20 2007
Copyright (c) 1999-2015, ADTRAN, Inc.
Platform: Total Access 908 (2nd Gen), part number 4212908L1
^MSerial number CFG1366002
Flash: 33554432 bytes  DRAM: 67108863 bytes

host1 uptime is 50 weeks, 5 days, 6 hours, 53 minutes, 18 seconds

^Mslot 0, DSP 1
  DSP software version: G2.R10.8.M.4b
  DSP hardware version: Freescale MSC7119
  Total channels: 30

System returned to ROM by External Hard Reset
Current system image file is "NONVOL:/9212916-2R100906.biz"
Boot system image file is "NONVOL:/9212916-2R100906.biz"
Primary system configuration file is "startup-config"
host1#show running-config
Building configuration...
!
!
! ADTRAN, Inc. OS version R10.9.6.E
```

**Comment by ytti** [comment]
is this still an issue? If so, can you provide remote access to sanitised device. I'd login from 91.198.120.1, you can supply authentication details to saku@ytti.fi

**Comment by thecityofguanyu** [comment]
@ytti 

I replied at the email address you listed.

**Comment by banananananananana** [comment]
So, I now have this exact issue... But what was the solution?

[link to relevant xkcd ](https://imgs.xkcd.com/comics/wisdom_of_the_ancients.png)

**Comment by ytti** [comment]
I've forgotten this, I should take a look, but won't have time this week or next due to traveling. Anyhow the problem is almost certainly prompt detection. Or using ```#send``` in model when prompt is expected (in which case ```#cmd``` should be used).

@thecityofguanyu @banananananananana please bug me in two weeks time if not resolved.

**Comment by thecityofguanyu** [comment]
@banananananananana 

I don't recall the exact details of when we were experiencing this issue, but I'm sure that it was indeed prompt related. Are you using Adtran AOS devices? Below is the `adtran.rb` model that is currently working for us.

```ruby
class ADTRAN < Oxidized::Model

  prompt /^.*[>#]$/
  comment  '! '

  cmd :all do |cfg|
    cfg.each_line.to_a[1..-2].join
  end

  cmd 'show version' do |cfg|
    comment cfg.lines.first
  end

  cmd 'show running-config' do |cfg|
    cfg = cfg.each_line.to_a[3..-1].join
    cfg
  end

  cfg :telnet do
    username /^Username:/
    password /^Password:/
  end

  cfg :telnet, :ssh do
    if vars :enable
      post_login do
        send "enable\n"
        cmd vars(:enable) + "\n"
      end
    else
      post_login do
        send "enable\n"
      end
    end
    post_login 'terminal length 0'
    pre_logout 'exit'
    pre_logout 'exit'
  end
end
```

**Comment by banananananananana** [comment]
I just did some proper troubleshooting, while keeping it as simple as possible.

I took one of the devices that is giving me problems, and put it in a new router.db. 

I made the output to a "file", and made ther interval at 30 seconds.

The ios.rb ONLY does a "show running-config"

--------
  cmd 'show running-config' do |cfg|
  cfg
  end
--------

It might run fine for 2-4 polls. On the fifth poll, the config is empty. On the 6th poll, the config returns.

When it gives me empty output, i've checked the debug log, and everything looks correct. It logs in, does a show run, and logs out. The complete config is in the debug log... But the output file is empty.

This is kinda driving me nuts.


gem list | grep oxid
oxidized (0.19.0)
oxidized-script (0.3.1)
oxidized-web (0.8.0)

**Comment by banananananananana** [comment]
Alright... So, the problem is related to telnet and the enable password, i think.

Me scenario: cisco devices with telnet access, with no enable password set.

If i set the enable password to "" in the config file, the output is empty...In spite of the debug log showing everything works as intended.

If i change the ios.rb file to do a "post_login 'enable' - then it works. But then again, some devices have the enable password set, so this wont work in the long run. sigh.

**Comment by sanjmonkey** [comment]
**TL;DR:** The problem is caused by **send "enable\n"** if your device doesn't need you to enable and you're using Telnet.


FWIW; I've observed and managed to replicate this behaviour.  After a bit of debugging I noticed the telnet command outputs were being read in but actually being processed against the next commands code block.  

eg; if you have something like;


```
class FOO < Oxidized::Model

  prompt /^([\w.@()-]+[#>]\s?)$/
  comment  '! '

  cmd 'show foo' do |cfg|
    cfg.gsub! /^Something I dont like in foo : [^\n]*\n/, ''
    comment cfg

  cmd 'show bar' do |cfg|
    cfg.gsub! /^Something I dont like in bar : [^\n]*\n/, ''
    comment cfg

  cfg :telnet do
    username /^Username:/
    password /^Password:/
  end

  cfg :telnet, :ssh do
    # preferred way to handle additional passwords
    if vars :enable
      post_login do
        send "enable\n"
        cmd vars(:enable)
      end
    end
    post_login 'terminal length 0'
    pre_logout 'exit'
  end

````

then the output of 'show foo' would be run against the 'show bar' code block.

**NOTE;**
* 'send' in this case causes Net::Telnet to just send data and not read in the output from the socket after send.
* 'cmd' will make Net::Telnet send the data, and trigger the slightly customised version of the read

If you **send "enable\n"** to a device which is already in this operational enabled mode, it does this:
```
switch#enable
switch#            <----------this is important
```

If you then **cmd password**, it would look like this (if you're box isn't doing domain-lookup)
```
switch#password
Translating "password"
% Unknown command or computer name, or unable to find computer address
switch#
```

Put those 2 together and you get:
```
switch#enable
switch#password
Translating "password"
% Unknown command or computer name, or unable to find computer address
switch#
```

Now because **cmd** is the only one which reads the buffer, the socket still has ```enable\nswitch#``` waiting to be read off it because send doesnt read it.  When cmd sends password, the command is sent and the readpartial code is triggered almost immediately after.  Because its so quick, he sees ```enable\nswitch#``` and thinks he's matched a prompt and the code moves on.

However, when the next commands are run, in the example file above: **post_login 'terminal length 0'**, the buffer isn't empty because the output from sending password has arrived in the meantime and looks like: ```password\nswitch#```.  And so on and so forth, thereby making every command run, but the output get matched against the following blocks code substitutions.

What you could do? (in order of preference)
* fix your devices.db to not include the enable var for these devices where you already land enabled...
* use another input method such as SSH ?
* look to pattern match your enable prompt specifically in your device model, by sending a newline eg:
```
  cfg :telnet, :ssh do
    # preferred way to handle additional passwords
    if vars :enable
      post_login do
        line = cmd ''
        enabled = true if line =~ /^(\n[\w.@()-]+#)$/
        if not enabled
          send "enable\n"
          cmd vars(:enable)
        end
      end
    end
    post_login 'terminal length 0'
    post_login 'terminal width 0'
    pre_logout 'exit'
  end
```
* put a time gap between the send data and the read off the buffer in telnet.rb (real hack) - around line 110

      begin
        sleep(2)
        c = @sock.readpartial(1024 * 1024)
`


Sandy

**Comment by ytti** [comment]
You always need to consume prompt. If you want to issue 'enable', but do not expect password prompt, then 'send' is incorrect.

```send``` does not expect anything. ```cmd``` expects prompt, or if another argument given, that.

If you _ALWAYS_ expect prompt after ```enable``` just change ```send``` to ```cmd```.  If you sometimes expect prompt, sometimes password prompt, then you need to be conditional.

Something like:
```
if vars(:enable) == True
  cmd "enable"
elsif vars(:enable)
  send "enable"
  cmd vars(:enable)
end
```

If enable is set to true, but not specific string, we expect regular prompt after 'enable', if enable is set to string, we expect nothing, but send enable password string next (as if there was enable password prompt). Otherwise, we set nothing.

Personally I don't get the point of enable at all, much less enable without password prompt.

**Comment by sanjmonkey** [comment]
Exactly, I imagine the OP was already enabled and had no reason to send enable again. 

Sandy

**Comment by wk** [comment]
If you continue working on this issue and Adtran support, please be aware that an Adtran model has been introduced in #1576, and ideally, further work should be toward extending that model with any workarounds your specific circumstances require. 

I will close this issue for now as it appears to be abandoned.
